_posts/2024-12-04-enable-pluggable-storage-in-opensearch-vectordb.md

```diff
@@ -0,0 +1,255 @@
+
+For more information about building a k-NN similarity search engine with OpenSearch, see our [documentation](https://opensearch.org/docs/latest/search-plugins/knn/index/).
+
+## Challenges with file-based APIs
+
+Native vector engines, such ass Faiss and NMSLIB, offer high performance and predictable latencies. However, their reliance on file-based APIs makes it difficult to integrate them with storage that is not based on file systems. 
+
+Lucene uses the Java-based **Directory** abstraction for reading and writing files, native engines tightly couple their operations to file-based I/O. The **Directory** class abstracts file storage, enabling operations like reading, writing, and managing file metadata across diverse storage systems. This abstraction allows the Lucene vector engine to store files independently of the underlying storage in OpenSearch.
+
+Unlike Lucene, native engines tightly couple their operations to file-based I/O. To address these limitations, we applied Lucene's principles to native engines. By abstracting the I/O layer, we eliminated the engines' tight coupling with specific file APIs. This enhancement enables integration with any OpenSearch directory implementation, making vector search compatible with a broader range of storage systems.
+
+## Introducing the loading layer
+
+Both Faiss and NMSLIB load graph-based vector indexes from storage into memory. During this process, they use `fread` to fetch the bytes needed to reconstruct the graph. 
+
+To improve flexibility, we replaced `fread` with a read interface. Faiss provides an **IOReader** interface for reading index data from various storage systems. For NMSLIB, we introduced a similar read interface called **NmslibIOReader**. These interfaces allow the native engines to read data through an abstraction layer, making it possible to integrate with OpenSearch's directory implementations.
+
+Because k-NN search is conducted after the graph is loaded into memory, this change does not impact average search performance. 
+
+The following diagram provides a high-level overview of the loading layer in native engines.
+
+![High-level overview of the loading layer in native engines](/assets/media/blog-images/2024-12-04-enable-pluggable-storage-in-opensearch-vectordb/loading_layer_high_level.png){:style="width: 100%; max-width: 800px;"}
+
+## Performance benchmarks
+
+The following sections present performance benchmark results.
+
+### Benchmarking environment
+
+We ran benchmarking tests in an environment with the following configuration.
+
+| OpenSearch version      | 2.18                  |
+| vCPUs                   | 48                    |
+| Physical memory         | 128 GB                |
+| Storage type            | Elastic Block Storage |
```

Comment:
By "Elastic Block Storage", do we mean "Amazon Elastic Block Store (Amazon EBS)"?

Reply:
@0ctopus13prime Could you confirm?

Reply:
Yes, it is. 
https://aws.amazon.com/ebs/